aggregator/src/aggregator/collection_job_driver.rs

```diff
@@ -224,64 +248,78 @@ where
             return Ok(());
         }
 
-        let (mut leader_aggregate_share, report_count, checksum) =
-            compute_aggregate_share::<SEED_SIZE, Q, A>(&task, &batch_aggregations)
-                .await
-                .map_err(|e| datastore::Error::User(e.into()))?;
-
-        vdaf.add_noise_to_agg_share(
-            &dp_strategy,
-            collection_job.aggregation_parameter(),
-            &mut leader_aggregate_share,
-            report_count.try_into()?,
-        )
-        .map_err(Error::DifferentialPrivacy)?;
-
-        // Send an aggregate share request to the helper.
-        let request = AggregateShareReq::<Q>::new(
-            BatchSelector::new(collection_job.batch_identifier().clone()),
-            collection_job.aggregation_parameter().get_encoded()?,
-            report_count,
-            checksum,
-        );
-
-        let resp_bytes = send_request_to_helper(
-            &self.http_client,
-            self.backoff.clone(),
-            Method::POST,
-            task.aggregate_shares_uri()?.ok_or_else(|| {
-                Error::InvalidConfiguration("task is not leader and has no aggregate share URI")
-            })?,
-            AGGREGATE_SHARES_ROUTE,
-            Some(RequestBody {
-                content_type: AggregateShareReq::<TimeInterval>::MEDIA_TYPE,
-                body: Bytes::from(request.get_encoded()?),
-            }),
-            // The only way a task wouldn't have an aggregator auth token in it is in the taskprov
-            // case, and Janus never acts as the leader with taskprov enabled.
-            task.aggregator_auth_token()
-                .ok_or_else(|| Error::InvalidConfiguration("no aggregator auth token in task"))?,
-            &self.metrics.http_request_duration_histogram,
-        )
-        .await?;
-
-        // Store the helper aggregate share in the datastore so that a later request to a collect
-        // job URI can serve it up.
         let collection_job = Arc::new(
-            collection_job.with_state(CollectionJobState::Finished {
-                report_count,
-                encrypted_helper_aggregate_share: AggregateShare::get_decoded(&resp_bytes)?
-                    .encrypted_aggregate_share()
-                    .clone(),
-                leader_aggregate_share,
-            }),
+            if let Some(finished_collection_job) = finished_collection_job {
```

Comment:
Note that writing the logic this way will skip contacting the Helper on a duplicated collection job. The only downside (AFAICT) is that if the Helper's keys have changed, this doesn't give the Helper another chance to encrypt its aggregate share.

I can change this, but it's a bit tricky: we'd need to change scrubbed `batch_aggregations` to keep their checksum around, since a finished collection job doesn't store the checksum. (And I don't want to introduce storing the checksum in the collection job for this purpose, since it will rarely be used & it would not be a backportable change.)

Reply:
Right you are, thanks.

Reply:
(I'm actually curious whether omitting the request to the Helper is _allowed_ based on the current text of DAP, but not enough to not implement this performance optimization.)